docs/puzzles/day5.md

```diff
@@ -2,12 +2,206 @@
+which are the start and end of the vent. I delegate the parsing of positions
+further into the apply metohd of Position:
+
+```scala
+object Point:
+  def apply(str: String) =
+    str.split(",") match
+      case Array(start, end) => new Point(start.trim.toInt, end.trim.toInt)
+      case _ =>
+        throw new java.lang.IllegalArgumentException(s"Wrong point input $str")
+```
+
+With all the above code we now have a really nicely modeled input.
+
+### Hydrothermal Venture: The Orthogonal Pleasures.
+
+Let's use the input to find all the dangerous spots on the map in the first part
+of the puzzle. We can use a mutable map with default values for that:
+
+```scala
+val map = mutable.Map[Point, Int]().withDefaultValue(0)
+```
+
+This way we can easily update all the points that have a vent defined with as
+imple method:
+
+```scala
+  def update(p: Point) =
+    val current = map(p)
+    map.update(p, current + 1)
+```
+
+Last thing to consider is how to actually add those points. For that we need to
+define proper ranges:
+
+```scala
+def rangex =
+  val stepx = if vent.end.x > vent.start.x then 1 else -1
+  vent.start.x.to(vent.end.x, stepx)
+def rangey =
+  val stepy = if vent.end.y > vent.start.y then 1 else -1
+  vent.start.y.to(vent.end.y, stepy)
+```
+
+We need to make sure that those range are increasing or decreasing based on the
+fact that the start might be located before the end.
+
+With those ranges defined we can just iterate and add all the points:
+
+```scala
+// vent is horizontal
+if vent.start.x == vent.end.x then
+  for py <- rangey do update(Point(vent.start.x, py))
+// vent is vertical
+else if vent.start.y == vent.end.y then
+  for px <- rangex do update(Point(px, vent.start.y))
+```
+
+and add the end just count them all up:
+
+```scala
+map.count { case (_, v) => v > 1 }
+```
+
+Altogether the solution looks like this:
+
+```scala
+def findDangerousPoints(vents: Seq[Vent]) =
+  val map = mutable.Map[Point, Int]().withDefaultValue(0)
+  def update(p: Point) =
+    val current = map(p)
+    map.update(p, current + 1)
+
+  for vent <- vents do
+    def rangex =
+      val stepx = if vent.end.x > vent.start.x then 1 else -1
+      vent.start.x.to(vent.end.x, stepx)
+    def rangey =
+      val stepy = if vent.end.y > vent.start.y then 1 else -1
+      vent.start.y.to(vent.end.y, stepy)
+    // vent is horizontal
+    if vent.start.x == vent.end.x then
+      for py <- rangey do update(Point(vent.start.x, py))
+    // vent is vertical
+    else if vent.start.y == vent.end.y
+      for px <- rangex do update(Point(px, vent.start.y))
+  end for
+
+  map.count { case (_, v) => v > 1 }
+end findDangerousPoints
+```
+
+### Hydrothermal Venture 2: The Diagonal Menace
+
+For the second part of the puzzle we need to take into account also the diagnal
+vents. Fortunately, they can only have an angle of 45 degrees, which means when
+iterating over ranges both x and y positions increment by 1. So we can add
```

Comment:
I find this sentence hard to read:
```suggestion
vents. Fortunately, they can only have an angle of 45 degrees, which means that
both x and y positions increment by 1 at each step of the range. So we can add
```

Reply:
I was writing it a bit in a hurry :sweat: